geo/geogfn: implement DWithin, Distance for spheroids

otan opened the pull request:
geo/geogfn: implement DWithin, Distance for spheroids

This commit introduces calculating Distance and DWithin for spheroids.
This is rolled out by hand as the S2 primitives were not available to do
this for us. Some minor optimizations have been made, but there are some
TODOs left in there for a future date.

It is important to note this is not the true minimum distance, but is
the minimum distance that is in line with the PostGIS definition. This
is commented on the `minDistance` function.

Also made a minor speed boost improvement for geographiclib, only
initializing the spheroid once.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48124)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> Reviewed 8 of 9 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48124)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geogfn/distance.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6ARDvo4D3tra9Tf-N_:-M6ARDvo4D3tra9Tf-Na:b-b8647y) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L65)):*
> ```Go
> 	for _, aRegion := range aRegions {
> 		for _, bRegion := range bRegions {
> 			// We cannot make the min distance updater be global across all collections,
> ```

I didn't understand why it can't be shared across all collections. The only state it has is related to the min distance found so far and we are early exiting so it doesn't seem like we necessarily compare all pairs.
___
*[pkg/geo/geogfn/distance.go, line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6APqBP71pBXEx3poRs:-M6APqBQ2vANWUYUAk_H:b-mfidu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L119)):*
> ```Go
> // NOTE: this is technically incorrect, as we are judging the minimum distance as
> // the minimum spheroid distance by using the minimum sphere distance to determine
> // the closest two edges. This is in line with PostGIS, which are are aiming
> ```

which we are aiming
___
*[pkg/geo/geogfn/distance.go, line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AQGhI7u2zbzsf1ZJs:-M6AQGhI7u2zbzsf1ZJt:b-a8v93f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L136)):*
> ```Go
> 		if u.stopAfterLE > 0 && sphereDistance <= u.stopAfterLE {
> 			return true
> 		}
> ```

simpler to write a single if-statement instead of two.
```
if u.minD <= u.stopAfterLE {
  return true
}
```
___
*[pkg/geo/geogfn/distance.go, line 261 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AZdqf7cQXvBGnVrsi:-M6AZdqf7cQXvBGnVrsj:bsrp4qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L261)):*
> ```Go
> 			} {
> 				// Check the vertex against the ends of the edges.
> 				if u.updateMinDistance(toCheck.vertex, toCheck.edge.V0) || u.updateMinDistance(toCheck.vertex, toCheck.edge.V1) {
> ```

nit: long line
___
*[pkg/geo/geogfn/distance.go, line 306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AYa46Bi7Svvt401vP:-M6AYa46Bi7Svvt401vQ:blypqwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L306)):*
> ```Go
> 	// Edge is really inefficient for polygons, so do the iterations.
> 	for _, aLoop := range a.Loops() {
> 		for _, bLoop := range b.Loops() {
> ```

So we will also compare inner loops to inner loops?
Is that meaningful or necessary?
___
*[pkg/geo/geogfn/geographiclib.go, line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AUcZ7ArKEPxY8Q8vI:-M6AUcZ7ArKEPxY8Q8vJ:b-xww9zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/geographiclib.go#L19)):*
> ```Go
> 
> // spheroidInverse returns the s12 (meter) component of spheroid.Inverse from s2 Points.
> func spheroidInverse(s geographiclib.Spheroid, a s2.Point, b s2.Point) float64 {
> ```

since we are only returning the length of the geodesic that connects these points, we should rename this method.
Also, whenever I see the inverse function being called I need to go search for what it means since this is so specific to Geodesy -- could you add a comment at the call to `s.Inverse` that this is solving the "geodetic inverse" problem (which is easy to google) and optionally add a link https://en.wikipedia.org/wiki/Geodesy#Geodetic_problems
___
*[pkg/geo/geogfn/s2.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AXcFtEmmqDilUmwPy:-M6AXcFtEmmqDilUmwPz:bj7b4s6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/s2.go#L53)):*
> ```Go
> // and a bool of whether it lies in the segment of the edge.
> // This operation is done on a sphere.
> func maybeClosestPointToEdge(edge s2.Edge, point s2.Point) (s2.Point, bool) {
> ```

can you add some commentary on what this algorithm is doing and why.
___
*[pkg/geo/geographiclib/geographiclib.go, line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AU2gP68VzvxYz1TEv:-M6AU2gP68VzvxYz1TEw:b-oaspjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geographiclib/geographiclib.go#L17)):*
> ```Go
> // #cgo LDFLAGS: -lm
> //
> // #include "geodesic.h"
> ```

can you clarify why the switch from geographiclib to geodesic?
and why we need the `spheroidInverse` function in geographiclib.go


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48124)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/distance.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6ARDvo4D3tra9Tf-N_:-M6B-1pp3j5BJu0y3Kiu:b-j6uzk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L65)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I didn't understand why it can't be shared across all collections. The only state it has is related to the min distance found so far and we are early exiting so it doesn't seem like we necessarily compare all pairs.
</blockquote></details>

Because the way postgis works is that it finds the minimum distance is done *per collection* using a sphere to calculate the two closest point, then returning a spheroid distance.
If we used all collections, there may be a separate two closest points that may have a longer sphere distance but shorter spheroid distance, giving us a different result to postgis.
___
*[pkg/geo/geogfn/distance.go, line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6APqBP71pBXEx3poRs:-M6B-0if-vanDMvKUQ7W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L119)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

which we are aiming
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AQGhI7u2zbzsf1ZJs:-M6Azx7z55ndctqrIXKz:b-ksp31c) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L136)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

simpler to write a single if-statement instead of two.
```
if u.minD <= u.stopAfterLE {
  return true
}
```
</blockquote></details>

Done (relic of spheroid math)
___
*[pkg/geo/geogfn/distance.go, line 261 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AZdqf7cQXvBGnVrsi:-M6AzkH3BB4dWUG6-lrq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L261)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: long line
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/geographiclib.go, line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AUcZ7ArKEPxY8Q8vI:-M6AzQOb0wICs7XAd7zH:b-wlrbpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/geographiclib.go#L19)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

since we are only returning the length of the geodesic that connects these points, we should rename this method.
Also, whenever I see the inverse function being called I need to go search for what it means since this is so specific to Geodesy -- could you add a comment at the call to `s.Inverse` that this is solving the "geodetic inverse" problem (which is easy to google) and optionally add a link https://en.wikipedia.org/wiki/Geodesy#Geodetic_problems
</blockquote></details>

it already does, I added a link.
renamed this one.
___
*[pkg/geo/geogfn/s2.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AXcFtEmmqDilUmwPy:-M6AyLpQBkM3WVO8nveC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/s2.go#L53)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you add some commentary on what this algorithm is doing and why.
</blockquote></details>

Done.
___
*[pkg/geo/geographiclib/geographiclib.go, line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AU2gP68VzvxYz1TEv:-M6Ap0q59A-cEKW9WH6G:b-kdcama) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geographiclib/geographiclib.go#L17)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you clarify why the switch from geographiclib to geodesic?
and why we need the `spheroidInverse` function in geographiclib.go
</blockquote></details>

geographiclib.h was a wrapper around geodesic.h, and i don't need it to do these based operations.
spheroidInverse makes it such that I don't need to discard the az1, az2 arguments, which we'll need later for ST_Azimuth.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48124)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/distance.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6ARDvo4D3tra9Tf-N_:-M6Gcy2j0FKC6Bge23df:ber05fs) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L65)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Because the way postgis works is that it finds the minimum distance is done *per collection* using a sphere to calculate the two closest point, then returning a spheroid distance.
If we used all collections, there may be a separate two closest points that may have a longer sphere distance but shorter spheroid distance, giving us a different result to postgis.
</blockquote></details>

Then this deservers a longer comment that first describes the PostGIS behavior, then states that `spheroidMinDistanceUpdater` finds the closest pair of points on two `s2.Regions` using a sphere and that it only uses spheroid for calculating the distance between that pair.
___
*[pkg/geo/geogfn/distance.go, line 281 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GkkWh9a3yCC8iQ76O:-M6GkkWh9a3yCC8iQ76P:b-6x72ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/distance.go#L281)):*
> ```Go
> func (u *spheroidMinDistanceUpdater) onPolylineToPolygon(a *s2.Polyline, b *s2.Polygon) bool {
> 	// If the first point of a is outside the exterior ring, then we just need to check the
> 	// exterior.
> ```

This needs a longer comment.
If we know there is at least one point outside the exterior ring, either
- all points are outside and we need to compare only against the exterior ring
- if some points are inside, the line must cross the exterior -- again we only need to compare against the exterior
___
*[pkg/geo/geogfn/distance.go, line 292 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GlysS7xJOIaH8Cop7:-M6GlysS7xJOIaH8Cop8:bpeluq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/distance.go#L292)):*
> ```Go
> 			return true
> 		}
> 		for _, linePoint := range *a {
> ```

this for loop could be made conditional on !insideHole
___
*[pkg/geo/geogfn/distance.go, line 318 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GmVq7CzwPB6yshhuD:-M6GmVq7CzwPB6yshhuE:b-atk0mb) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/distance.go#L318)):*
> ```Go
> 
> 	// If the first point of both outer loops are not within the other,
> 	// we only have to compare outer rings.
> ```

This could use a longer comment. Something like
If we know there is at least one point on both outer loops that are outside the outside outer loop we have one of two cases:
- The outer loops do not intersect. We can find the closest points by comparing only the outer loops.
- The outer loops intersect and the minimum distance is 0. Again we can discover this intersection by comparing only the outer loops.
___
*[pkg/geo/geogfn/distance.go, line 323 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6Gn2NZ4YSU8cQ_BpIF:-M6Gn2NZ4YSU8cQ_BpIG:bqfgxck) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/distance.go#L323)):*
> ```Go
> 	}
> 
> 	// If the first point of A is inside a hole of B, check A exterior against B hole.
> ```

Longer comment like.
Since if this point is inside a hole, either (a) the A is inside the hole and the closest point can be found by comparing A's outer loop and this hole in B, (b) A intersects this hole and the distance is zero, which can also be found by comparing A's outer loop and this hole in B. 
___
*[pkg/geo/geogfn/s2.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GiFt1316h-2n_hHU8:-M6GiFt1316h-2n_hHU9:b-vq3mq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/s2.go#L58)):*
> ```Go
> 	// the normal and the point will intersect with the great circle represented
> 	// by the given edge.
> 	normal := edge.V0.Vector.Cross(edge.V1.Vector).Normalize()
> ```

I don't understand why we are projecting on to the normal and what the "it" is in "subtract it".
Do you have a reference to the algorithm, so I can read it?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48124)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geogfn/distance.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6ARDvo4D3tra9Tf-N_:-M6GswpQ7_PbNSE9ws8Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L65)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Then this deservers a longer comment that first describes the PostGIS behavior, then states that `spheroidMinDistanceUpdater` finds the closest pair of points on two `s2.Regions` using a sphere and that it only uses spheroid for calculating the distance between that pair.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AYa46Bi7Svvt401vP:-M6GvbAr4eZqL45kpy3q:b-h9x3e1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L306)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

So we will also compare inner loops to inner loops?
Is that meaningful or necessary?
</blockquote></details>

inner loops to exterior rings only.
we never need to compare the interior ring.
___
*[pkg/geo/geogfn/distance.go, line 281 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GkkWh9a3yCC8iQ76O:-M6Gu79LCv-1oOdVYedG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/distance.go#L281)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This needs a longer comment.
If we know there is at least one point outside the exterior ring, either
- all points are outside and we need to compare only against the exterior ring
- if some points are inside, the line must cross the exterior -- again we only need to compare against the exterior
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 292 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GlysS7xJOIaH8Cop7:-M6Gu7rZ0eaavCpJOHIL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/distance.go#L292)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this for loop could be made conditional on !insideHole
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 318 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GmVq7CzwPB6yshhuD:-M6GvZCq7LpQ7ZlWlCC_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/distance.go#L318)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This could use a longer comment. Something like
If we know there is at least one point on both outer loops that are outside the outside outer loop we have one of two cases:
- The outer loops do not intersect. We can find the closest points by comparing only the outer loops.
- The outer loops intersect and the minimum distance is 0. Again we can discover this intersection by comparing only the outer loops.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 323 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6Gn2NZ4YSU8cQ_BpIF:-M6GvaYmCe1qrdqTOij0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/distance.go#L323)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Longer comment like.
Since if this point is inside a hole, either (a) the A is inside the hole and the closest point can be found by comparing A's outer loop and this hole in B, (b) A intersects this hole and the distance is zero, which can also be found by comparing A's outer loop and this hole in B. 

</blockquote></details>

Done.
___
*[pkg/geo/geogfn/geographiclib.go, line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AUcZ7ArKEPxY8Q8vI:-M6GxZ1A7DngRK_jwf03:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/geographiclib.go#L19)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

it already does, I added a link.
renamed this one.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/s2.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GiFt1316h-2n_hHU8:-M6Gvl2O-ATE2Wro4vJM:bmol1hm) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/s2.go#L58)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't understand why we are projecting on to the normal and what the "it" is in "subtract it".
Do you have a reference to the algorithm, so I can read it?

</blockquote></details>

I've only seen code with no explanations.  I've added something which helped me understand it more, but if you have better maths parlance I'm happy to copy what you think :)
___
*[pkg/geo/geographiclib/geographiclib.go, line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AU2gP68VzvxYz1TEv:-M6GxZX852yRw58wvmh_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geographiclib/geographiclib.go#L17)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

geographiclib.h was a wrapper around geodesic.h, and i don't need it to do these based operations.
spheroidInverse makes it such that I don't need to discard the az1, az2 arguments, which we'll need later for ST_Azimuth.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48124)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/distance.go, line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VQ9lq6zTaKeaqeHDu:-M6VQ9lr4aARWHXH3vq6:b5zfed6) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L65)):*
> ```Go
> 	for _, aRegion := range aRegions {
> 		for _, bRegion := range bRegions {
> 			// The way PostGIS evaluated the distance between spheroid regions is to
> ```

s/evaluated/evaluates/
___
*[pkg/geo/geogfn/distance.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VQkKtB_INO8mT9E-N:-M6VQkKtB_INO8mT9E-O:b4nd8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L79)):*
> ```Go
> 			// the minimum spheroid distance by using the minimum sphere distance to determine
> 			// the closest two edges. This is in line with PostGIS, which we are aiming
> 			// to be compatible with. See distance_test.go for examples of "truer" values.
> ```

I think this comment should be hoisted about the two for loops since it is also describing the overall algorithm. I found it a bit hard to read since it gives two different examples, and the latter "this is" is a bit unclear and it doesn't explain the behavior of `minDistanceUpdater`. How about something like the following:

We have two sets of regions. PostGIS defines the distance between such sets of regions as follows: compute the min of the pair-wise distance between the cross-product of these sets, where the pair-wise distance is computed as:
- Compute the two closest points between the pairs of regions, assuming a sphere.
- Compute the spheroid distance between the two closest points.
This is technically incorrect, since it is possible that the two closest points on the spheroid are different than the two closest points on the sphere. See distance_test.go for examples of the "truer" distance values. Since we aim to be compatible with PostGIS we adopt the same approach.

The spheroidMinDistanceUpdater is responsible for computing the distance between a pair of regions. It uses S2 to compute the closest points on the sphere. When done, spheroidMinDistanceUpater.minDistance() returns the spheroid distance between these points.
___
*[pkg/geo/geogfn/distance.go, line 298 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VTzNaClx-CyJTlV0_:-M6VTzNaClx-CyJTlV0a:bcc0eyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L298)):*
> ```Go
> 	}
> 
> 	// Now we are guaranteed that there is at least one point is inside the exterior ring.
> ```

... one point inside ...
___
*[pkg/geo/geogfn/distance.go, line 299 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VUB4f3CKED5UpUkvW:-M6VUB4f3CKED5UpUkvX:b-1p5n2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L299)):*
> ```Go
> 
> 	// Now we are guaranteed that there is at least one point is inside the exterior ring.
> 	// Keep track if any lines are inside a hole.
> ```

Alternative to this last sentence:

For a polygon with no holes, the fact that there is a point inside the exterior ring would imply that the distance is zero. However, when there are holes, it is possible that the distance is non-zero if polyline-a is completely contained inside a hole. We iterate over the holes and
- compute the distance between the hole and polyline-a.
- if polyline-a is within minD, we can immediately return.
- if polyline-a does not intersect the hole but there is at least one point inside the hole, we can immediately return since all points must be inside that hole (since distance must not have been zero, otherwise we would have returned in the previous step).
___
*[pkg/geo/geogfn/distance.go, line 308 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VhcPp7G2k_RPRl3P8:-M6VhcPp7G2k_RPRl3P9:brqcwb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L308)):*
> ```Go
> 			if hole.ContainsPoint(linePoint) {
> 				insideHole = true
> 			}
> ```

based on my suggested comment, I think we can also exit from this inner loop as soon as `insideHole = true`.

And I think we can remove `insideHole` altogether by hoisting the return statement here.
___
*[pkg/geo/geogfn/distance.go, line 334 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6ViTtf5njcGBsGRsV-:-M6ViTtf5njcGBsGRsV0:b-9bokb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L334)):*
> ```Go
> 	bFirstPoint := b.Loop(0).Vertex(0)
> 
> 	// If there is at least one point on the the exterior ring of B that is also the exterior ring
> ```

s/also/outside/
___
*[pkg/geo/geogfn/distance.go, line 335 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VigWZAeKfEpOlASIE:-M6VigWZAeKfEpOlASIF:b-3kd4ot) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L335)):*
> ```Go
> 
> 	// If there is at least one point on the the exterior ring of B that is also the exterior ring
> 	// of A, then we have one of these two cases:
> ```

// of A, and vice versa, ...

without that we could have A contained in B, so this reasoning would not apply.
___
*[pkg/geo/geogfn/distance.go, line 340 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6ViuxE-TQADRYzT2Sc:-M6ViuxE-TQADRYzT2Sd:b900ih2) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L340)):*
> ```Go
> 	// * The exterior rings of A and B never meet. This distance can always be found
> 	//   by only comparing the exterior rings.
> 	// This applies if one point on the exterior ring of A is on the exterior ring of B.
> ```

I don't see anything being added by the sentence "This applies ... ring of B", so suggest dropping it.
___
*[pkg/geo/geogfn/s2.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GiFt1316h-2n_hHU8:-M6Vl5bY8oiSmUlkCvsn:bnizb9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/s2.go#L58)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I've only seen code with no explanations.  I've added something which helped me understand it more, but if you have better maths parlance I'm happy to copy what you think :)
</blockquote></details>

Thanks -- I'll take a look, but it doesn't need to hold back this PR from being merged.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48124)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geogfn/distance.go, line 306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6AYa46Bi7Svvt401vP:-M6XtfVA-F7fpZVybAxG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9c28fa1e1851bc4707fbb0a68389c46ee59b2a/pkg/geo/geogfn/distance.go#L306)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

inner loops to exterior rings only.
we never need to compare the interior ring.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 65 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VQ9lq6zTaKeaqeHDu:-M6XtwzS4dMZAUb7D1Df:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L65)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/evaluated/evaluates/
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 79 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VQkKtB_INO8mT9E-N:-M6XyPoI1elE_P8TZUiq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L79)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think this comment should be hoisted about the two for loops since it is also describing the overall algorithm. I found it a bit hard to read since it gives two different examples, and the latter "this is" is a bit unclear and it doesn't explain the behavior of `minDistanceUpdater`. How about something like the following:

We have two sets of regions. PostGIS defines the distance between such sets of regions as follows: compute the min of the pair-wise distance between the cross-product of these sets, where the pair-wise distance is computed as:
- Compute the two closest points between the pairs of regions, assuming a sphere.
- Compute the spheroid distance between the two closest points.
This is technically incorrect, since it is possible that the two closest points on the spheroid are different than the two closest points on the sphere. See distance_test.go for examples of the "truer" distance values. Since we aim to be compatible with PostGIS we adopt the same approach.

The spheroidMinDistanceUpdater is responsible for computing the distance between a pair of regions. It uses S2 to compute the closest points on the sphere. When done, spheroidMinDistanceUpater.minDistance() returns the spheroid distance between these points.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 298 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VTzNaClx-CyJTlV0_:-M6XtB_-8fl1Lxota_iI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L298)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

... one point inside ...
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 299 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VUB4f3CKED5UpUkvW:-M6XvS-40Tu5wT-19HhZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L299)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Alternative to this last sentence:

For a polygon with no holes, the fact that there is a point inside the exterior ring would imply that the distance is zero. However, when there are holes, it is possible that the distance is non-zero if polyline-a is completely contained inside a hole. We iterate over the holes and
- compute the distance between the hole and polyline-a.
- if polyline-a is within minD, we can immediately return.
- if polyline-a does not intersect the hole but there is at least one point inside the hole, we can immediately return since all points must be inside that hole (since distance must not have been zero, otherwise we would have returned in the previous step).
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 308 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VhcPp7G2k_RPRl3P8:-M6XuVff1Qrw4WydDi4T:b1svo9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L308)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

based on my suggested comment, I think we can also exit from this inner loop as soon as `insideHole = true`.

And I think we can remove `insideHole` altogether by hoisting the return statement here.
</blockquote></details>

We cannot do that, as we still need to calculate the minimum distance.
___
*[pkg/geo/geogfn/distance.go, line 334 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6ViTtf5njcGBsGRsV-:-M6XtP80EQcu3oVlxvhV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L334)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/also/outside/
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 335 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VigWZAeKfEpOlASIE:-M6XvpKn1GdH6S54qJKR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L335)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

// of A, and vice versa, ...

without that we could have A contained in B, so this reasoning would not apply.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 340 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6ViuxE-TQADRYzT2Sc:-M6XtecfEK1IIz3HtT3m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L340)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't see anything being added by the sentence "This applies ... ring of B", so suggest dropping it.
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/s2.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GiFt1316h-2n_hHU8:-M6Xtgyk3fy8bbGSsdtk:b-5ag6wf) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/s2.go#L58)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Thanks -- I'll take a look, but it doesn't need to hold back this PR from being merged.
</blockquote></details>

It makes sense to me now but requires explaining math principles (what x-product and dot product do) which i think is too much.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geogfn/distance.go, line 308 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VhcPp7G2k_RPRl3P8:-M6XySXU8C_Dk96bwQLA:b36l9ah) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L308)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

We cannot do that, as we still need to calculate the minimum distance.
</blockquote></details>

oops, nvm


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 2 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48124)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/distance.go, line 335 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VigWZAeKfEpOlASIE:-M6_2v1GBWKnLljubiqa:bocb2l1) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L335)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I don't see this change.
___
*[pkg/geo/geogfn/distance.go, line 60 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6_1fPQCFqbtiENa-Ll:-M6_1fPQCFqbtiENa-Lm:bbiyb1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9865a5e6a093b0689d89b520c9c0a6361575e5d2/pkg/geo/geogfn/distance.go#L60)):*
> ```Go
> //
> // !!! SURPRISING BEHAVIOR WARNING !!!
> // The way PostGIS evaluates the distance between spheroid regions is the done
> ```

nit:
PostGIS evaluates the distance between spheroid regions by computing ...
___
*[pkg/geo/geogfn/distance.go, line 61 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6_1kwoEOa71XINFDAj:-M6_1kwoEOa71XINFDAk:bxgs3bj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9865a5e6a093b0689d89b520c9c0a6361575e5d2/pkg/geo/geogfn/distance.go#L61)):*
> ```Go
> // !!! SURPRISING BEHAVIOR WARNING !!!
> // The way PostGIS evaluates the distance between spheroid regions is the done
> // by compututing the min of the pair-wise distance between the cross-product of
> ```

computing
___
*[pkg/geo/geogfn/distance.go, line 71 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6_1w8762wCGhjFr5Pc:-M6_1w8762wCGhjFr5Pd:b-o8uwn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9865a5e6a093b0689d89b520c9c0a6361575e5d2/pkg/geo/geogfn/distance.go#L71)):*
> ```Go
> // See distance_test.go for examples of the "truer" distance values.
> // Since we aim to be compatible with PostGIS we adopt the same approach.
> // the closest two edges.
> ```

"the closest two edges." seems leftover from some previous revision.
___
*[pkg/geo/geogfn/s2.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6GiFt1316h-2n_hHU8:-M6_B7ab1EHhKSJW8se5:blfj3sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/77822cc78f325b16c5a48b68fb495bdcdfa276a5/pkg/geo/geogfn/s2.go#L58)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

It makes sense to me now but requires explaining math principles (what x-product and dot product do) which i think is too much.
</blockquote></details>

It's these operations on a sphere that I find it harder to reason about. But the explanation is good enough. I'll dig deeper after the PR is merged.


<!-- Sent from Reviewable.io -->

otan committed:
geo/geogfn: implement DWithin, Distance for spheroids

This commit introduces calculating Distance and DWithin for spheroids.
This is rolled out by hand as the S2 primitives were not available to do
this for us. Some minor optimizations have been made, but there are some
TODOs left in there for a future date.

It is important to note this is not the true minimum distance, but is
the minimum distance that is in line with the PostGIS definition. This
is commented on the `minDistance` function.

Also made a minor speed boost improvement for geographiclib, only
initializing the spheroid once.

Release note: None

otan commented:
___
*[pkg/geo/geogfn/distance.go, line 335 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-M6VigWZAeKfEpOlASIE:-M6_JdNh9aV4bPFg3rOx:b-xtbh1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/59dd4afaaf696833aa5f14dd2a17e22fce6edb47/pkg/geo/geogfn/distance.go#L335)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't see this change.
</blockquote></details>

Woops.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48124#-:-M6fC68iBhL0LbmpulNm:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)



<!-- Sent from Reviewable.io -->

otan commented:
tftr!

bors r=sumeerbhola
